### PR TITLE
gh-109598: make PyComplex_RealAsDouble/ImagAsDouble use __complex__

### PR DESCRIPTION
**Pull request by skirpichev**
PyComplex_ImagAsDouble() also will not silently return 0.0 for non-complex types anymore.  Instead we try to call PyFloat_AsDouble() and return 0.0 only if this call is successful.

* Issue: gh-109598

----
:books: Documentation preview :books:: https://cpython-previews--109647.org.readthedocs.build/

### CODE REVIEW

File: Objects/complexobject.c [hunk 1]
```diff
@@ -256,26 +256,51 @@ PyComplex_FromDoubles(double real, double imag)
     return PyComplex_FromCComplex(c);
 }
 
+static PyObject * try_complex_special_method(PyObject *);
+
 double
 PyComplex_RealAsDouble(PyObject *op)
 {
+    double real = -1.0;
```
Comment (Ogb-Bassey) [inline Objects/complexobject.c:264]: I am thinking about whether "real" and "imag" below should return something else by default if op is not a valid object. Say something like NaN, given that -1.0 can be a valid real or imaginary part

Comment (skirpichev) [reply Objects/complexobject.c:264]: ``1`` or ``1.0`` is a more common for C-API functions choice, if they return int/long/double, see e.g. ``PyLong_AsLong()`` or ``PyLong_AsDouble()``.   This convention was also extended to ``PyComplex_AsCComplex()``.   And it was used before in ``PyComplex_RealAsDouble()``, so your suggestion will break backward compatibility.

**Comment by skirpichev** [comment]
CC @mdickinson as suggested by https://devguide.python.org/experts/

**Comment by skirpichev** [comment]
Will factor out PyComplex C API tests in #111591.  Meanwhile, lets convert this to draft...

**Comment by skirpichev** [comment]
This is ready for review again, much smaller patch.

**Comment by skirpichev** [comment]
https://discuss.python.org/t/43004/

**Review by gvanrossum — APPROVED** [review]
LGTM. Very thorough.

**Review by encukou — APPROVED** [review]

**Pull request merged**